src/lib/Types/Core.hs

```diff
@@ -211,6 +212,12 @@ data PiType  (n::S) where
 data DepPairType (n::S) where
   DepPairType :: Binder n l -> Type  l -> DepPairType n
 
+data Projection
+  = UnwrapCompoundNewtype  -- Unwrap TypeCon, record or variant
+  | UnwrapBaseNewtype      -- Unwrap Fin or Nat
+  | ProjectProduct Int
```

Comment:
I assume this last case includes user-defined "newtypes", i.e. ADTs with a single unary constructor?

Reply:
No! The only projection valid for such an ADT is `UnwrapCompoundNewtype`, which returns a value of its representation type. For a unary ADT its representation is the payload, so that's the only thing you need to do.

Reply:
I see. So to project out the `Int` from a `data MyPair = MkMyPair Int Float` you'd use `[UnwrapCompoundNewType, ProjectProduct 0]`, and to project the `Int` from `MyNewtype = MkMyNewtype Int` you'd just use `[UnwrapCompoundNewtype]`? At some point in the past, we would have used `[0]` in both cases, but I suppose this change already happened with your newtype patch, just with integer indices instead of `Projection`.

All sounds good. Maybe add a comment to remind me that `UnwrapCompoundNewtype` unwraps both newtype-like ADTs and ordinary ADTs.

Reply:
The lengths of projection lists are the same as they were before! While `[0]` extracted an `Int` from `MyNewtype` (and now is `[UnwrapCompoundNewtype]`), you needed `[0, 0]` to get an `Int` out of `MkMyPair` (which now is `[ProjectProduct 0, UnwrapCompoundNewtype]`). This patch basically distinguishes the `0`s used for newtype unwrapping from the zeros used for product projections, but doesn't change the number of zeros you need.